core/src/main/java/org/geysermc/geyser/command/CommandRegistry.java

```diff
@@ -0,0 +1,333 @@
+                "geyser.commands.exthelp.desc",
+                "geyser.command.exthelp." + id,
+                extension.rootCommand(),
+                entry.getValue()));
+        }
+
+        // wait for the right moment (depends on the platform) to register permissions
+        geyser.eventBus().subscribe(new GeyserEventRegistrar(this), GeyserRegisterPermissionsEvent.class, this::onRegisterPermissions);
+    }
+
+    @NotNull
+    public CommandManager<GeyserCommandSource> cloud() {
+        return cloud;
+    }
+
+    @NotNull
+    public Map<String, Command> commands() {
+        return Collections.unmodifiableMap(this.commands);
+    }
+
+    @NotNull
+    public Map<Extension, Map<String, Command>> extensionCommands() {
+        return Collections.unmodifiableMap(this.extensionCommands);
+    }
+
+    public void clear() {
+        this.commands.clear();
+        this.extensionCommands.clear();
+        this.permissionDefaults.clear();
+    }
+
+    /**
+     * For internal Geyser commands
+     */
+    public void registerBuiltInCommand(GeyserCommand command) {
+        register(command, this.commands);
+    }
+
+    public void registerExtensionCommand(@NonNull Extension extension, @NonNull GeyserCommand command) {
+        register(command, this.extensionCommands.computeIfAbsent(extension, e -> new HashMap<>()));
+    }
+
+    private void register(GeyserCommand command, Map<String, Command> commands) {
+        command.register(cloud);
+
+        commands.put(command.name(), command);
+        geyser.getLogger().debug(GeyserLocale.getLocaleStringLog("geyser.commands.registered", command.name()));
+
+        for (String alias : command.aliases()) {
+            commands.put(alias, command);
+        }
+
+        if (!command.permission().isBlank() && command.permissionDefault() != null) {
+            permissionDefaults.put(command.permission(), command.permissionDefault());
+        }
+    }
+
+    private void onRegisterPermissions(GeyserRegisterPermissionsEvent event) {
+        for (Map.Entry<String, TriState> permission : permissionDefaults.entrySet()) {
+            event.register(permission.getKey(), permission.getValue());
+        }
+    }
+
+    /**
+     * Returns the description of the given command
+     *
+     * @param command the root command node
+     * @param locale the ideal locale that the description should be in
+     * @return a description if found, otherwise an empty string. The locale is not guaranteed.
+     */
+    @NonNull
+    public String description(@NonNull String command, @NonNull String locale) {
+        if (command.equals(GeyserCommand.DEFAULT_ROOT_COMMAND)) {
+            return GeyserLocale.getPlayerLocaleString("geyser.command.geyser_desc", locale);
+        }
+
+        Extension extension = extensionRootCommands.get(command);
+        if (extension != null) {
+            return GeyserLocale.getPlayerLocaleString("geyser.command.extension_desc", locale, extension.description().name());
```

Comment:
epic stuff, but may need to changed a tad if https://github.com/GeyserMC/languages/pull/115/files#r1327153024 is adapted